### ISSUE TITLE
ThreadSanitizer reports data race in `SSL_do_handshake`

### ISSUE THREAD

**Issue by oerdnj** [openssl/openssl#25202]
Hey, this doesn't have simple reproducer and ThreadSanitizer is confused and doesn't produce proper backtraces, but it has been happening for a several ~months~ years now in our CI (since April 2022). Why it sometimes happen and sometime is does not is probably related to the timing (and overall system stress).

So, ThreadSanitizer produces following report:
```
WARNING: ThreadSanitizer: data race 
  Read of size 8 at 0x000000000001 by thread T0001:
    #0 memcmp <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 ASN1_STRING_cmp build_shared/../crypto/asn1/asn1_lib.c:392:17 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #2 tls_do_bio lib/isc/netmgr/tlsstream.c:586:10 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #3 tls_readcb lib/isc/netmgr/tlsstream.c:840:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #4 isc___nm_readcb lib/isc/netmgr/netmgr.c:1859:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #5 isc__nm_readcb lib/isc/netmgr/netmgr.c:1874:3
    #6 isc__nm_tcp_read_cb lib/isc/netmgr/tcp.c:778:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #7 uv__read /usr/src/libuv-v1.48.0/src/unix/stream.c:1143:7 (BuildId: 7c913c4e0ce5027e1de4e434dcc0b6f749c36e45)
    #8 thread_body lib/isc/thread.c:85:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #9 thread_run lib/isc/thread.c:100:14

  Previous write of size 8 at 0x000000000001 by thread T0002 (mutexes: write M0001):
    #0 memcpy <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 memcpy /usr/include/x86_64-linux-gnu/bits/string_fortified.h:29:10 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #2 ASN1_STRING_set build_shared/../crypto/asn1/asn1_lib.c:318:9
    #3 tls_do_bio lib/isc/netmgr/tlsstream.c:586:10 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #4 tls_readcb lib/isc/netmgr/tlsstream.c:840:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #5 isc___nm_readcb lib/isc/netmgr/netmgr.c:1859:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #6 isc__nm_readcb lib/isc/netmgr/netmgr.c:1874:3
    #7 isc__nm_tcp_read_cb lib/isc/netmgr/tcp.c:778:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #8 uv__read /usr/src/libuv-v1.48.0/src/unix/stream.c:1143:7 (BuildId: 7c913c4e0ce5027e1de4e434dcc0b6f749c36e45)
    #9 thread_body lib/isc/thread.c:85:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #10 thread_run lib/isc/thread.c:100:14

  Location is heap block of size 21 at 0x000000000001 allocated by thread T0002:
    #0 malloc <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 mallocx lib/isc/./jemalloc_shim.h:67:14 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #2 mem_get lib/isc/mem.c:303:8
    #3 isc__mem_allocate lib/isc/mem.c:757:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #4 isc__mem_reallocate lib/isc/mem.c:805:13
    #5 isc__tls_realloc_ex lib/isc/tls.c:83:10 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #6 ASN1_STRING_set build_shared/../crypto/asn1/asn1_lib.c:308:21 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #7 tls_do_bio lib/isc/netmgr/tlsstream.c:586:10 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #8 tls_readcb lib/isc/netmgr/tlsstream.c:840:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #9 isc___nm_readcb lib/isc/netmgr/netmgr.c:1859:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #10 isc__nm_readcb lib/isc/netmgr/netmgr.c:1874:3
    #11 isc__nm_tcp_read_cb lib/isc/netmgr/tcp.c:778:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #12 uv__read /usr/src/libuv-v1.48.0/src/unix/stream.c:1143:7 (BuildId: 7c913c4e0ce5027e1de4e434dcc0b6f749c36e45)
    #13 thread_body lib/isc/thread.c:85:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #14 thread_run lib/isc/thread.c:100:14

  Mutex M0001 (0x000000000017) created at:
    #0 pthread_rwlock_init <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 CRYPTO_THREAD_lock_new build_shared/../crypto/threads_pthread.c:55:9 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #2 listenelt_create lib/ns/listenlist.c:86:15 (BuildId: 05070b7e65de6b70ed1ee897b6d0936b36db3f6a)
    #3 ns_listenelt_create lib/ns/listenlist.c:222:9 (BuildId: 05070b7e65de6b70ed1ee897b6d0936b36db3f6a)
    #4 listenelt_fromconfig bin/named/server.c:11377:3 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #5 listenlist_fromconfig bin/named/server.c:11093:12
    #6 load_configuration bin/named/server.c:8816:13 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #7 run_server bin/named/server.c:9968:2 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #8 isc__async_cb lib/isc/async.c:111:3 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #9 uv__async_io /usr/src/libuv-v1.48.0/src/unix/async.c:176:5 (BuildId: 7c913c4e0ce5027e1de4e434dcc0b6f749c36e45)
    #10 thread_body lib/isc/thread.c:85:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #11 isc_thread_main lib/isc/thread.c:116:2
    #12 isc_loopmgr_run lib/isc/loop.c:462:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #13 main bin/named/main.c:1548:2 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)

  Thread T0001 'isc-loop-0003' (running) created by main thread at:
    #0 pthread_create <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 isc_thread_create lib/isc/thread.c:139:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #2 isc_loopmgr_run lib/isc/loop.c:456:3 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #3 main bin/named/main.c:1548:2 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)

  Thread T0002 'isc-loop-0002' (running) created by main thread at:
    #0 pthread_create <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 isc_thread_create lib/isc/thread.c:139:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #2 isc_loopmgr_run lib/isc/loop.c:456:3 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #3 main bin/named/main.c:1548:2 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)

SUMMARY: ThreadSanitizer: data race (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998) in memcmp
```

As I said, I have no idea why it strips the backtraces, but it does 🤷 , so I am adding comments below:

```
 Mutex M0001 (0x000000000017) created at:
    #0 pthread_rwlock_init <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 CRYPTO_THREAD_lock_new build_shared/../crypto/threads_pthread.c:55:9 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #2 listenelt_create lib/ns/listenlist.c:86:15 (BuildId: 05070b7e65de6b70ed1ee897b6d0936b36db3f6a)
```

In here, we create a cert store:
```c
isc_result_t
isc_tls_cert_store_create(const char *ca_bundle_filename,
                          isc_tls_cert_store_t **pstore) {
        int ret = 0;
        isc_tls_cert_store_t *store = NULL;
        REQUIRE(pstore != NULL && *pstore == NULL);

        store = X509_STORE_new();
        if (store == NULL) {
                goto error;
        }

        /* Let's treat empty string as the default (system wide) store */
        if (ca_bundle_filename != NULL && *ca_bundle_filename == '\0') {
                ca_bundle_filename = NULL;
        }

        if (ca_bundle_filename == NULL) {
                ret = X509_STORE_set_default_paths(store);
        } else {
                ret = X509_STORE_load_locations(store, ca_bundle_filename,
                                                NULL);
        }

        if (ret == 0) {
                goto error;
        }

        *pstore = store;
        return (ISC_R_SUCCESS);

error:
        ERR_clear_error();
        if (store != NULL) {
                X509_STORE_free(store);
        }
        return (ISC_R_FAILURE);
}
```

This happens when a "listener" for an interface is created.  This creates an internal mutex in the OpenSSL, right?

The next part is about write:
```
  Previous write of size 8 at 0x000000000001 by thread T0002 (mutexes: write M0001):
    #0 memcpy <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 memcpy /usr/include/x86_64-linux-gnu/bits/string_fortified.h:29:10 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #2 ASN1_STRING_set build_shared/../crypto/asn1/asn1_lib.c:318:9
    #3 tls_do_bio lib/isc/netmgr/tlsstream.c:586:10 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
```

The `ASN1_STRING_set` is protected by the internal mutex (created in the previous step), and the code in `#3` is actually this function:
```c
static int
tls_try_handshake(isc_nmsocket_t *sock, isc_result_t *presult) {
        REQUIRE(sock->tlsstream.state == TLS_HANDSHAKE);

        if (SSL_is_init_finished(sock->tlsstream.tls) == 1) {
                return (0);
        }

        int rv = SSL_do_handshake(sock->tlsstream.tls);
        if (rv == 1) {
                isc_nmhandle_t *tlshandle = NULL;
                isc_result_t result = ISC_R_SUCCESS;

                REQUIRE(sock->statichandle == NULL);
                INSIST(SSL_is_init_finished(sock->tlsstream.tls) == 1);

                isc__nmsocket_log_tls_session_reuse(sock, sock->tlsstream.tls);
                tlshandle = isc__nmhandle_get(sock, &sock->peer, &sock->iface);
                tls_read_stop(sock);

                if (isc__nm_closing(sock->worker)) {
                        result = ISC_R_SHUTTINGDOWN;
                }

                if (sock->tlsstream.server) {
                        /*
                         * The listening sockets are now closed from outer
                         * to inner order, which means that this function
                         * will never be called when the outer socket has
                         * stopped listening.
                         *
                         * Also see 'isc__nmsocket_stop()' - the function used
                         * to shut down the listening TLS socket - for more
                         * details.
                         */
                        if (result == ISC_R_SUCCESS) {
                                result = sock->accept_cb(tlshandle, result,
                                                         sock->accept_cbarg);
                        }
                } else {
                        tls_call_connect_cb(sock, tlshandle, result);
                }
                isc_nmhandle_detach(&tlshandle);
                sock->tlsstream.state = TLS_IO;

                if (presult != NULL) {
                        *presult = result;
                }
        }

        return (rv);
}
```

As far as I can tell with debugger in one hand and breakpoints in other, the problematic part comes from `SSL_do_handshake()`, but since the backtraces from ThreadSanitizer are incomplete, I can't really tell which `d2i_SSL_SESSION_ex()` is the problematic one in the `construct_stateless_ticket`, but all the local backtraces point to that function (there's about 11 calls to `ASN1_STRING_set()` during a single connection):

The first six are same from frame 7 to frame 32:
```
#0  ASN1_STRING_set (str=0x720800026700, _data=0x722800014fac, len_in=2) at crypto/asn1/asn1_lib.c:289
#1  0x00007ffff6f05a7a in asn1_ex_c2i (pval=0x7234000063f8,
    cont=0x722800014fac "\023\002\004 ^/\332Xo\2113\227\271|\377t\345s\270\217\335F\r\177\271,\036\356\360\200\026\232o&r\247\0040y\033f\265\242Y\353\020\304:Hs\210\245zPj\236I\361\206+\321Z\241\334e\244\017J5\313\352\353&l]fT\031\241C/N7,\351Q\241\006\002\004f\275\374\025\242\004\002\002\034 \244\026\004\024\353\034\233\243\330v7\202", len=2, utype=4, free_cont=0x7ffffffd715d "", it=0x7ffff758fc50 <ASN1_OCTET_STRING_it.local_it>) at crypto/asn1/tasn_dec.c:941
#2  0x00007ffff6f03a30 in asn1_d2i_ex_primitive (pval=0x7234000063f8, in=0x7ffffffd7728, inlen=145, it=0x7ffff758fc50 <ASN1_OCTET_STRING_it.local_it>, tag=4, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8) at crypto/asn1/tasn_dec.c:818
#3  0x00007ffff6f01a03 in asn1_item_embed_d2i (pval=0x7234000063f8, in=0x7ffffffd7728, len=145, it=0x7ffff758fc50 <ASN1_OCTET_STRING_it.local_it>, tag=-1, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8, depth=2, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:217
#4  0x00007ffff6f04c05 in asn1_template_noexp_d2i (val=0x7234000063f8, in=0x7ffffffd7c20, len=145, tt=0x7ffff78f1ac0 <SSL_SESSION_ASN1_seq_tt+80>, opt=0 '\000', ctx=0x7ffffffd7db8, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:682
#5  0x00007ffff6f03167 in asn1_template_ex_d2i (val=0x7234000063f8, in=0x7ffffffd7c20, inlen=145, tt=0x7ffff78f1ac0 <SSL_SESSION_ASN1_seq_tt+80>, opt=0 '\000', ctx=0x7ffffffd7db8, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:558
#6  0x00007ffff6f028b1 in asn1_item_embed_d2i (pval=0x7ffffffd7db0, in=0x7ffffffd83d8, len=145, it=0x7ffff78f1a30 <SSL_SESSION_ASN1_it.local_it>, tag=16, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:422
#7  0x00007ffff6f0143e in asn1_item_ex_d2i_intern (pval=0x7ffffffd7db0, in=0x7ffffffd83d8, len=155, it=0x7ffff78f1a30 <SSL_SESSION_ASN1_it.local_it>, tag=-1, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:118
#8  0x00007ffff6f01555 in ASN1_item_d2i_ex (pval=0x7ffffffd7db0, in=0x7ffffffd83d8, len=155, it=0x7ffff78f1a30 <SSL_SESSION_ASN1_it.local_it>, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:144
#9  0x00007ffff6f0160e in ASN1_item_d2i (pval=0x0, in=0x7ffffffd83d8, len=155, it=0x7ffff78f1a30 <SSL_SESSION_ASN1_it.local_it>) at crypto/asn1/tasn_dec.c:154
#10 0x00007ffff7769726 in d2i_SSL_SESSION_ASN1 (a=0x0, in=0x7ffffffd83d8, len=155) at ssl/ssl_asn1.c:84
#11 0x00007ffff77682a4 in d2i_SSL_SESSION_ex (a=0x0, pp=0x7ffffffd86d0, length=155, libctx=0x0, propq=0x0) at ssl/ssl_asn1.c:272
#12 0x00007ffff78b9391 in construct_stateless_ticket (s=0x728800000000, pkt=0x7ffffffd8c50, age_add=1255546958, tick_nonce=0x7ffffffd8a50 "") at ssl/statem/statem_srvr.c:3995
#13 0x00007ffff78ae4c3 in tls_construct_new_session_ticket (s=0x728800000000, pkt=0x7ffffffd8c50) at ssl/statem/statem_srvr.c:4287
#14 0x00007ffff787f6a5 in write_state_machine (s=0x728800000000) at ssl/statem/statem.c:894
#15 0x00007ffff787dd62 in state_machine (s=0x728800000000, server=1) at ssl/statem/statem.c:487
#16 0x00007ffff787e05e in ossl_statem_accept (s=0x728800000000) at ssl/statem/statem.c:307
#17 0x00007ffff77893c1 in SSL_do_handshake (s=0x728800000000) at ssl/ssl_lib.c:4746
#18 0x00007ffff7f135fb in tls_try_handshake (sock=0x726800004e00, presult=0x7fffffff93e0) at netmgr/tlsstream.c:458
#19 0x00007ffff7f0ce7b in tls_do_bio (sock=0x726800004e00, received_data=0x7fffffff9530, send_data=0x0, finish=false) at netmgr/tlsstream.c:586
#20 0x00007ffff7f15859 in tls_readcb (handle=0x723c00003930, result=ISC_R_SUCCESS, region=0x7fffffff9530, cbarg=0x726800004e00) at netmgr/tlsstream.c:840
#21 0x00007ffff7ee8c2d in isc___nm_readcb (arg=0x725400001e00) at netmgr/netmgr.c:1859
#22 0x00007ffff7ee5b8e in isc__nm_readcb (sock=0x726800005a00, uvreq=0x725400001e00, eresult=ISC_R_SUCCESS, async=false) at netmgr/netmgr.c:1874
#23 0x00007ffff7f07ea3 in isc__nm_tcp_read_cb (stream=0x726800005cc8, nread=314, buf=0x7fffffff9a08) at netmgr/tcp.c:778
#24 0x00007ffff76ea883 in uv__read (stream=0x726800005cc8) at src/unix/stream.c:1143
#25 0x00007ffff76e64df in uv__stream_io (loop=0x727000000820, w=0x726800005d50, events=1) at src/unix/stream.c:1203
#26 0x00007ffff76f895d in uv__io_poll (loop=0x727000000820, timeout=2070) at src/unix/linux.c:1528
#27 0x00007ffff76c5f3f in uv_run (loop=0x727000000820, mode=UV_RUN_DEFAULT) at src/unix/core.c:448
#28 0x00007ffff7f4e744 in loop_thread (arg=0x727000000800) at loop.c:288
#29 0x00007ffff7f78974 in thread_body (wrap=0x7208000194e0) at thread.c:85
#30 0x00007ffff7f78864 in isc_thread_main (func=0x7ffff7f4e510 <loop_thread>, arg=0x727000000800) at thread.c:116
#31 0x00007ffff7f4e3c4 in isc_loopmgr_run (loopmgr=0x7230000006c0) at loop.c:462
#32 0x0000555555662579 in main (argc=6, argv=0x7fffffffde58) at main.c:1548
```

And the difference between traces 01 ... 06 are all similar to:
```patch
--- /tmp/01	2024-08-15 15:06:19.419039850 +0200
+++ /tmp/02	2024-08-15 15:06:24.567039965 +0200
@@ -1,11 +1,11 @@
-#0  ASN1_STRING_set (str=0x720800026700, _data=0x722800014fac, len_in=2) at crypto/asn1/asn1_lib.c:289
-#1  0x00007ffff6f05a7a in asn1_ex_c2i (pval=0x7234000063f8,
-    cont=0x722800014fac "\023\002\004 ^/\332Xo\2113\227\271|\377t\345s\270\217\335F\r\177\271,\036\356\360\200\026\232o&r\247\0040y\033f\265\242Y\353\020\304:Hs\210\245zPj\236I\361\206+\321Z\241\334e\244\017J5\313\352\353&l]fT\031\241C/N7,\351Q\241\006\002\004f\275\374\025\242\004\002\002\034 \244\026\004\024\353\034\233\243\330v7\202", len=2, utype=4, free_cont=0x7ffffffd715d "", it=0x7ffff758fc50 <ASN1_OCTET_STRING_it.local_it>) at crypto/asn1/tasn_dec.c:941
-#2  0x00007ffff6f03a30 in asn1_d2i_ex_primitive (pval=0x7234000063f8, in=0x7ffffffd7728, inlen=145, it=0x7ffff758fc50 <ASN1_OCTET_STRING_it.local_it>, tag=4, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8) at crypto/asn1/tasn_dec.c:818
-#3  0x00007ffff6f01a03 in asn1_item_embed_d2i (pval=0x7234000063f8, in=0x7ffffffd7728, len=145, it=0x7ffff758fc50 <ASN1_OCTET_STRING_it.local_it>, tag=-1, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8, depth=2, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:217
-#4  0x00007ffff6f04c05 in asn1_template_noexp_d2i (val=0x7234000063f8, in=0x7ffffffd7c20, len=145, tt=0x7ffff78f1ac0 <SSL_SESSION_ASN1_seq_tt+80>, opt=0 '\000', ctx=0x7ffffffd7db8, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:682
-#5  0x00007ffff6f03167 in asn1_template_ex_d2i (val=0x7234000063f8, in=0x7ffffffd7c20, inlen=145, tt=0x7ffff78f1ac0 <SSL_SESSION_ASN1_seq_tt+80>, opt=0 '\000', ctx=0x7ffffffd7db8, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:558
-#6  0x00007ffff6f028b1 in asn1_item_embed_d2i (pval=0x7ffffffd7db0, in=0x7ffffffd83d8, len=145, it=0x7ffff78f1a30 <SSL_SESSION_ASN1_it.local_it>, tag=16, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:422
+#0  ASN1_STRING_set (str=0x7208000266e0, _data=0x722800014fb0, len_in=32) at crypto/asn1/asn1_lib.c:289
+#1  0x00007ffff6f05a7a in asn1_ex_c2i (pval=0x723400006410,
+    cont=0x722800014fb0 "^/\332Xo\2113\227\271|\377t\345s\270\217\335F\r\177\271,\036\356\360\200\026\232o&r\247\0040y\033f\265\242Y\353\020\304:Hs\210\245zPj\236I\361\206+\321Z\241\334e\244\017J5\313\352\353&l]fT\031\241C/N7,\351Q\241\006\002\004f\275\374\025\242\004\002\002\034 \244\026\004\024\353\034\233\243\330v7\202", len=32, utype=4, free_cont=0x7ffffffd715d "", it=0x7ffff758fc50 <ASN1_OCTET_STRING_it.local_it>) at crypto/asn1/tasn_dec.c:941
+#2  0x00007ffff6f03a30 in asn1_d2i_ex_primitive (pval=0x723400006410, in=0x7ffffffd7728, inlen=141, it=0x7ffff758fc50 <ASN1_OCTET_STRING_it.local_it>, tag=4, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8) at crypto/asn1/tasn_dec.c:818
+#3  0x00007ffff6f01a03 in asn1_item_embed_d2i (pval=0x723400006410, in=0x7ffffffd7728, len=141, it=0x7ffff758fc50 <ASN1_OCTET_STRING_it.local_it>, tag=-1, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8, depth=2, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:217
+#4  0x00007ffff6f04c05 in asn1_template_noexp_d2i (val=0x723400006410, in=0x7ffffffd7c20, len=141, tt=0x7ffff78f1ae8 <SSL_SESSION_ASN1_seq_tt+120>, opt=0 '\000', ctx=0x7ffffffd7db8, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:682
+#5  0x00007ffff6f03167 in asn1_template_ex_d2i (val=0x723400006410, in=0x7ffffffd7c20, inlen=141, tt=0x7ffff78f1ae8 <SSL_SESSION_ASN1_seq_tt+120>, opt=0 '\000', ctx=0x7ffffffd7db8, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:558
+#6  0x00007ffff6f028b1 in asn1_item_embed_d2i (pval=0x7ffffffd7db0, in=0x7ffffffd83d8, len=141, it=0x7ffff78f1a30 <SSL_SESSION_ASN1_it.local_it>, tag=16, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:422
 #7  0x00007ffff6f0143e in asn1_item_ex_d2i_intern (pval=0x7ffffffd7db0, in=0x7ffffffd83d8, len=155, it=0x7ffff78f1a30 <SSL_SESSION_ASN1_it.local_it>, tag=-1, aclass=0, opt=0 '\000', ctx=0x7ffffffd7db8, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:118
 #8  0x00007ffff6f01555 in ASN1_item_d2i_ex (pval=0x7ffffffd7db0, in=0x7ffffffd83d8, len=155, it=0x7ffff78f1a30 <SSL_SESSION_ASN1_it.local_it>, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:144
 #9  0x00007ffff6f0160e in ASN1_item_d2i (pval=0x0, in=0x7ffffffd83d8, len=155, it=0x7ffff78f1a30 <SSL_SESSION_ASN1_it.local_it>) at crypto/asn1/tasn_dec.c:154
```

The difference between trace 06 and 07 starts here:
```
-#12 0x00007ffff78b9391 in construct_stateless_ticket (s=0x728800000000, pkt=0x7ffffffd8c50, age_add=1255546958, tick_nonce=0x7ffffffd8a50 "") at ssl/statem/statem_srvr.c:3995
+#12 0x00007ffff78b9391 in construct_stateless_ticket (s=0x728800000000, pkt=0x7ffffffd8c50, age_add=1092764678, tick_nonce=0x7ffffffd8a50 "") at ssl/statem/statem_srvr.c:3995
```

And the rest (07 ... 11) differs only in the frames 0 to 6 again.

I don't think there is anything that we can do ourselves to prevent this kind of data race, but I am happy to either help solve this issue in OpenSSL by providing more info or solving this on our side (BIND 9).

**Comment by nhorman** [comment]
I'm not sure what you mean by "stripping backtraces", the backtraces in your initial comment seem pretty complete to me, unless I'm missing something (entirely possible).

The initial backtraces that you provide in your initial comment (annotated):
```
WARNING: ThreadSanitizer: data race 
  Read of size 8 at 0x000000000001 by thread T0001:
    #0 memcmp <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 ASN1_STRING_cmp build_shared/../crypto/asn1/asn1_lib.c:392:17 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #2 tls_do_bio lib/isc/netmgr/tlsstream.c:586:10 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #3 tls_readcb lib/isc/netmgr/tlsstream.c:840:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #4 isc___nm_readcb lib/isc/netmgr/netmgr.c:1859:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #5 isc__nm_readcb lib/isc/netmgr/netmgr.c:1874:3
    #6 isc__nm_tcp_read_cb lib/isc/netmgr/tcp.c:778:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #7 uv__read /usr/src/libuv-v1.48.0/src/unix/stream.c:1143:7 (BuildId: 7c913c4e0ce5027e1de4e434dcc0b6f749c36e45)
    #8 thread_body lib/isc/thread.c:85:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #9 thread_run lib/isc/thread.c:100:14
```
This says that you read data in your call to ASN1_string_cmp
```
  Previous write of size 8 at 0x000000000001 by thread T0002 (mutexes: write M0001):
    #0 memcpy <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 memcpy /usr/include/x86_64-linux-gnu/bits/string_fortified.h:29:10 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #2 ASN1_STRING_set build_shared/../crypto/asn1/asn1_lib.c:318:9
    #3 tls_do_bio lib/isc/netmgr/tlsstream.c:586:10 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #4 tls_readcb lib/isc/netmgr/tlsstream.c:840:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #5 isc___nm_readcb lib/isc/netmgr/netmgr.c:1859:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #6 isc__nm_readcb lib/isc/netmgr/netmgr.c:1874:3
    #7 isc__nm_tcp_read_cb lib/isc/netmgr/tcp.c:778:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #8 uv__read /usr/src/libuv-v1.48.0/src/unix/stream.c:1143:7 (BuildId: 7c913c4e0ce5027e1de4e434dcc0b6f749c36e45)
    #9 thread_body lib/isc/thread.c:85:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #10 thread_run lib/isc/thread.c:100:14
```
This says that you previously wrote that same location without the benefit of a lock to serialize that access

```
  Location is heap block of size 21 at 0x000000000001 allocated by thread T0002:
    #0 malloc <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 mallocx lib/isc/./jemalloc_shim.h:67:14 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #2 mem_get lib/isc/mem.c:303:8
    #3 isc__mem_allocate lib/isc/mem.c:757:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #4 isc__mem_reallocate lib/isc/mem.c:805:13
    #5 isc__tls_realloc_ex lib/isc/tls.c:83:10 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #6 ASN1_STRING_set build_shared/../crypto/asn1/asn1_lib.c:308:21 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #7 tls_do_bio lib/isc/netmgr/tlsstream.c:586:10 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #8 tls_readcb lib/isc/netmgr/tlsstream.c:840:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #9 isc___nm_readcb lib/isc/netmgr/netmgr.c:1859:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #10 isc__nm_readcb lib/isc/netmgr/netmgr.c:1874:3
    #11 isc__nm_tcp_read_cb lib/isc/netmgr/tcp.c:778:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #12 uv__read /usr/src/libuv-v1.48.0/src/unix/stream.c:1143:7 (BuildId: 7c913c4e0ce5027e1de4e434dcc0b6f749c36e45)
    #13 thread_body lib/isc/thread.c:85:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #14 thread_run lib/isc/thread.c:100:14
```
this is reporting on the location where that RAW warning was alocated

```
  Mutex M0001 (0x000000000017) created at:
    #0 pthread_rwlock_init <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 CRYPTO_THREAD_lock_new build_shared/../crypto/threads_pthread.c:55:9 (BuildId: 774aef226915a125987ff800900a0b2ffb88a1cc)
    #2 listenelt_create lib/ns/listenlist.c:86:15 (BuildId: 05070b7e65de6b70ed1ee897b6d0936b36db3f6a)
    #3 ns_listenelt_create lib/ns/listenlist.c:222:9 (BuildId: 05070b7e65de6b70ed1ee897b6d0936b36db3f6a)
    #4 listenelt_fromconfig bin/named/server.c:11377:3 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #5 listenlist_fromconfig bin/named/server.c:11093:12
    #6 load_configuration bin/named/server.c:8816:13 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #7 run_server bin/named/server.c:9968:2 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #8 isc__async_cb lib/isc/async.c:111:3 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #9 uv__async_io /usr/src/libuv-v1.48.0/src/unix/async.c:176:5 (BuildId: 7c913c4e0ce5027e1de4e434dcc0b6f749c36e45)
    #10 thread_body lib/isc/thread.c:85:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #11 isc_thread_main lib/isc/thread.c:116:2
    #12 isc_loopmgr_run lib/isc/loop.c:462:2 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #13 main bin/named/main.c:1548:2 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
```
This is an announcement of the creation of a mutex (unrelated to the RAW error above)
```

  Thread T0001 'isc-loop-0003' (running) created by main thread at:
    #0 pthread_create <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 isc_thread_create lib/isc/thread.c:139:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #2 isc_loopmgr_run lib/isc/loop.c:456:3 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #3 main bin/named/main.c:1548:2 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)

  Thread T0002 'isc-loop-0002' (running) created by main thread at:
    #0 pthread_create <null> (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)
    #1 isc_thread_create lib/isc/thread.c:139:8 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #2 isc_loopmgr_run lib/isc/loop.c:456:3 (BuildId: 3576c9bdc1ef7a8e340576f0c1d7a0811bdd2cb5)
    #3 main bin/named/main.c:1548:2 (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998)

SUMMARY: ThreadSanitizer: data race (BuildId: f50900da42d53c3527b708e518eb8ea8f6260998) in memcmp
```
This is an announcement of the creation of two threads (again not seemingly related to the RAW error)

So what this in total says to me is that your function tls_do_bio is being executed by two threads in parallel, which is reading/writing the same data without the benefit of any serialization between them

Can you post the contents of your tls_do_bio function?

**Comment by oerdnj** [comment]
I can't tell ... because the I can't tell what object the `ASN1_STRING_cmp()` nor `ASN1_STRING_set()` operate on when the TSAN issue happens.

I do however have a local copy of OpenSSL compiled with TSAN, so I can insert artifical delays and fprintf debugs in places.

**Comment by davidben** [comment]
It'd be a bit strange for the `ASN1_STRING_set` call in the backtrace you listed to be the culprit because it acts on a freshly-allocated `SSL_SESSION` object. Unless OpenSSL has done something really strange under the hood reusing objects, or maybe there's a use-after-free somewhere in your code that's confusing TSan.

I don't suppose you have a complete stack trace from the debugger from `ASN1_STRING_cmp`?

I've also noticed that your incomplete stack traces pretty consistently lose all stack frames from OpenSSL. So whatever is the cause of TSan's trimmed stack traces, it has something to do with how you built OpenSSL. I also noticed your OpenSSL build comes out of "build_shared". Given that...

1. Did you TSan-instrument just your project, or OpenSSL as well?
2. Sanitizers often interact weirdly with shared libraries. I don't know off-hand if TSan is one such case, but does it reproduce if you statically link OpenSSL?

**Comment by oerdnj** [comment]
> I don't suppose you have a complete stack trace from the debugger from ASN1_STRING_cmp?

I don't at the moment, but I can do that quite easily. Give me a second.

> I also noticed your OpenSSL build comes out of "build_shared".

Hmm, well spotted. 

> Did you TSan-instrument just your project, or OpenSSL as well?

We now have two images `gcc:tsan` and `clang:tsan`, the `gcc:tsan` uses self-compiled TSAN-instrumented OpenSSL. But alas, the ThreadSanitizer warning comes from `clang:tsan`.  I guess I'll have to do the same for the `clang:tsan` image to get the fully instrumented backtraces. Thanks that was helpful.

> Sanitizers often interact weirdly with shared libraries. I don't know off-hand if TSan is one such case, but does it reproduce if you statically link OpenSSL?

Let me fix the `clang:tsan` image first and then I can try this. But from my experience with ThreadSanitizer, it doesn't have any problems with shared libraries. But it works better if those are fully-instrumented.

The most problematic part is that this doesn't happen always, just often enough to be a nuisance.

**Comment by nhorman** [comment]
ok, that is strange.  I can't imagine that optimzation  has inlined that many functions.  What openssl version are you using, and is it possible to build openssl with -g -O0 -fno-inline-functions?

That should eliminate any odd inlining that might mess with your backtrace

**Comment by oerdnj** [comment]
I am now in the process of rebuilding the second image we use for `clang:test` testing and this should give us full backtraces from ThreadSanitizer when the issue happens again in our CI.  It will just need couple days for the issue to happen again (ideally, it won't and we will be done here).

**Comment by oerdnj** [comment]
> I don't suppose you have a complete stack trace from the debugger from ASN1_STRING_cmp?

I tried to get this from a simple reproducer - `named` running and it receives a single DNS-over-HTTPS query, and there was no `ASN1_STRING_cmp` invoked. So, this looks more complicated. Hopefully, the CI will produce more complete backtrace next time.

**Comment by davidben** [comment]
Oh did you mix GCC TSan and Clang TSan together? I've never tried it, but I would not have a priori expect the two to play well together.

**Comment by oerdnj** [comment]
Ok, so the work was fruitful: https://isc-projects.gitlab-pages.isc.org/-/bind9/-/jobs/4630680/artifacts/tsan/2e048f595d23a06b86f8183458051826045c6f2ce97cc9ba4155004f78ca9d72/2bfb79ff1bb5544530ca97f2b37451aa8ccdbdf4ebb992891ff31f9bb8440731.txt

This one is similar, but little bit different: https://isc-projects.gitlab-pages.isc.org/-/bind9/-/jobs/4630680/artifacts/tsan/49962f9941fb97c3284fe26789e9e289f5325abfde150b3a610b5fee4d6014b7/a4e90b4efbd94088a430941c0cde8d413d3d0906284796e4e104b1185eb937da.txt

All of the reports: https://gitlab.isc.org/isc-projects/bind9/-/jobs/4630680/artifacts/browse/tsan/

**Comment by mattcaswell** [comment]
This looks like it could be the issue recently fixed by https://github.com/openssl/openssl/pull/24782

**Comment by mattcaswell** [comment]
> This looks like it could be the issue recently fixed by https://github.com/openssl/openssl/pull/24782

That fix is in git for all 3.x branches but hasn't yet been incorporated into a patch release.

**Comment by oerdnj** [comment]
What a coincidence - that would be awesome. I could cherry-pick the patch and add it to our images next week and see whether I can torture the CI to produce something.

**Comment by oerdnj** [comment]
I have cherry-picked the patch on top of 3.3.1 and after 200 CI jobs it's clean.  I am going to run this couple more times, but I think this hit the nail on the head! Thank you

**Comment by oerdnj** [comment]
Looks like it was duplicate of #24782 after all. I don't know why TSAN was so confused about the call stack when OpenSSL was not instrumented, but the paths leading the reported data race from our instrumented code checks out.

There's about 400 CI jobs finished and not a single report about the previous issue after I applied the patch.

Even though this is a duplicate, the discovery that it has been already fixed and we need to use future patched versions (or apply the patch) in our CI was also tremendously helpful. Thank you.

Now only to fix the other zillion issue we have ;).
